### ISSUE TITLE
CLIP finetuning in YOLO World

### ISSUE THREAD

**Issue by toobatehreem** [ultralytics/ultralytics#17615]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

How can I finetune CLIP in YOLO World with a custom learning rate?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @toobatehreem, thank you for your interest in Ultralytics 🚀!

We recommend a visit to the [Docs](https://docs.ultralytics.com) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. Your question about finetuning CLIP might have helpful hints in our documentation.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us debug it. 

For your ❓ Question about finetuning CLIP, please share more details, such as the specific commands or configurations you're using. This information will aid the engineering team in assisting you better.

You can also connect with the vibrant Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://ultralytics.com/discord) 🎧. For more in-depth discussions, check out [Discourse](https://community.ultralytics.com). Or join conversations on our [Subreddit](https://reddit.com/r/ultralytics) to share knowledge with fellow users.

## Upgrade

Please ensure you are using the latest `ultralytics` package, along with all required dependencies. You can upgrade in a [**Python>=3.8**](https://www.python.org/) environment using:

```bash
pip install -U ultralytics
```

## Environments

YOLO models can run in the following environments pre-loaded with necessary dependencies ([CUDA](https://developer.nvidia.com/cuda), [CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), [PyTorch](https://pytorch.org/)):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are currently passing. CI tests aim to verify the correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu.

This response is automated, and an Ultralytics engineer will be with you shortly to assist further. 😊

**Comment by glenn-jocher** [comment]
@toobatehreem to fine-tune CLIP within YOLO-World using a custom learning rate, you can modify the training script to set your desired learning rate in the optimizer settings. Ensure you specify the learning rate parameter when initializing your optimizer in the training configuration. If you need further guidance, please refer to the [YOLO-World documentation](https://docs.ultralytics.com/models/yolo-world) for detailed instructions.

**Comment by toobatehreem** [comment]
> @toobatehreem to fine-tune CLIP within YOLO-World using a custom learning rate, you can modify the training script to set your desired learning rate in the optimizer settings. Ensure you specify the learning rate parameter when initializing your optimizer in the training configuration. If you need further guidance, please refer to the [YOLO-World documentation](https://docs.ultralytics.com/models/yolo-world) for detailed instructions.

I want to use a learning rate of 0.0002 for YOLO backbone and a learning rate of 0.01 for CLIP like it is done in original YOLO World, how can I set 2 different learning rates both the models?

**Comment by glenn-jocher** [comment]
@toobatehreem to set different learning rates for the YOLO backbone and CLIP in YOLO-World, you will need to configure separate parameter groups in your optimizer. This involves specifying different learning rates for each group in your training script. You can define these parameter groups when initializing the optimizer, typically within the model's training configuration file or script. For detailed guidance on configuring parameter groups, please refer to the [YOLO-World documentation](https://docs.ultralytics.com/models/yolo-world).

**Comment by toobatehreem** [comment]
> @toobatehreem to set different learning rates for the YOLO backbone and CLIP in YOLO-World, you will need to configure separate parameter groups in your optimizer. This involves specifying different learning rates for each group in your training script. You can define these parameter groups when initializing the optimizer, typically within the model's training configuration file or script. For detailed guidance on configuring parameter groups, please refer to the [YOLO-World documentation](https://docs.ultralytics.com/models/yolo-world).

The problem is that even after doing this, the grads for CLIP remain zero. Since the texts are passed to the CLIP, the tokens and features are generated once, and after that, the features are detached. I don't think setting the text model requires grad to True, and passing the params in the optimizer param group can fine-tune CLIP.

**Comment by glenn-jocher** [comment]
Thank you for the clarification. In YOLO-World, it's correct that CLIP text features are detached to prevent gradients from propagating. To enable fine-tuning for CLIP, ensure that the `requires_grad` attribute for the relevant CLIP parameters (e.g., text encoder) is explicitly set to `True` before adding them to the optimizer. Additionally, avoid detaching the features if you want gradients to flow during backward passes. Please review the CLIP and YOLO-World integration carefully and adjust as needed to fine-tune effectively. Further implementation details might require custom adjustments in the codebase.
